CORS-3855: Remove ARO build flag from installer

rna-afk opened the pull request:
Removing the ARO build flag from the installer to achieve the goal of moving ARO installs closer to the installer. Ideally, these ARO specific changes to the installs/installer need to be done either in the wrapper or hive.

sadasu commented:
/retest-required

zaneb commented on pkg/asset/installconfig/azure/validation.go:
```diff
@@ -711,18 +711,15 @@ func validateResourceGroup(client API, fieldPath *field.Path, platform *aztypes.
-				ids = ids[:2]
-			}
-			allErrs = append(allErrs, field.Invalid(fieldPath.Child("resourceGroupName"), platform.ResourceGroupName, fmt.Sprintf("resource group must be empty but it has %d resources like %s ...", l, strings.Join(ids, ", "))))
+	ids, err := client.ListResourceIDsByGroup(context.TODO(), platform.ResourceGroupName)
```
I don't think we can just enable this code unconditionally. Rather, we need to move it to somewhere that the ARO installer naturally doesn't run instead of having it behind an explicit flag.

A possible pattern might be to move this validation to the PlatformProvisionCheck asset, which doesn't run until the `create cluster` phase (which ARO doesn't use). @patrickdillon do you have any thoughts on that?

patrickdillon replied:
In general, validations should go here if they apply to both UPI & IPI and in PlatformProvisionCheck if IPI-only. I often find it very difficult to reason about this distinction...

Moving to PlatformProvisionCheck makes sense, but talking through it:

The risk we have by moving the validation is that somebody creates a UPI cluster in a resource group that contains other resources and then runs `destroy cluster` and wipes out the whole resource group.

Initially my thought was "nobody doing UPI is going to do that" but the result is pretty catastrophic, so that makes me hesitant...

Doesn't ARO set a special marker on the resource group (managed by or something?) Could we inspect the resource group and see if it contains that special ARO flag, and skip the validation just for them? LMK if that idea isn't clear.

patrickdillon replied:
Following up on my previous comment,  ARO does indeed set [ManagedBy](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.management.resourcemanager.models.resourcegroup.managedby?view=azure-dotnet-legacy) on the rg, but I believe the value is non-determinsitic and non-discoverable. 

Does ARO use the installer for destroy? If not, I would consider skipping this check if ManagedBy is set and disallowing destroy if it is set. 

If not,

rna-afk replied:
We do have an env variable `OPENSHIFT_INSTALLER_INVOKER` that is set to ARO

zaneb replied:
I am certain that ARO is not using destroy today, because it is not implemented in the wrapper.
It is using Hive, however, so I am not sure if there is a scenario in the future where it could end up using destroy? (I presume not though.)

I think the solution here is that `destroy cluster` needs to be sure not to do anything if `create cluster` has not started successfully (i.e. we cannot have created anything that needs to be destroyed). If that were the case then it would be safe to move the validation into the `create cluster` phase (instead of the `create manifests` phase). I think in practice successfully starting could correspond to the `PlatformProvisionCheck` asset existing in the asset store if we put the validations inside that asset. Currently we run `destroy cluster` as long as the file from the `cluster.Metadata` asset exists on disk (which is created by either `create ignition-configs` or `create cluster`). The fact that this works after only doing `create ignition-configs` even without doing `create cluster` seems to be an unintentional side-effect of #1070. That actually seems like a bug since if the user follows a UPI-like process (UPI-_like_ because `destroy cluster` implies a non-`none` platform) we can't actually know what their infra looks like to destroy it.

On the other ideas:

We've previously said that we don't want to condition behaviour on environment variables. Perhaps this would be a justifiable exception to that rule, but the road map has ARO eventually driving the install through an unmodified Hive, at which point we would no longer have a custom invoker (and in fact it has already been reported that it is not working, though I don't understand how). So we would likely only be kicking the can down the road.

The install-config contains an option for passing an existing ResourceGroup rather than the installer creating one (this is the mechanism by which we plan to allow ARO to set ManagedBy). Presumably if people are using this then they may also be setting ManagedBy on the ResourceGroup themselves as well, so conditioning on that would change the behaviour for them. If the installer creates the ResourceGroup itself then it should by definition be empty, so the purpose of this check is to make sure one that is passed in is empty.

An option is to still check if the ResourceGroup is empty, but only warn instead of erroring, and store whether or not it was empty in the Metadata. Then upon destroy we would raise an error if it hadn't been empty to start with. That wouldn't affect any existing users, since up to now they've had to provide an empty ResourceGroup. But it is a little late - you'd only be informed that you have this thing that requires manual deletion after it has already been created.

rna-afk replied:
Added a commit with the fix.

rna-afk commented:
@patrickdillon I'm working on fixing this validation but there's no reason to have a client check at this part.
https://github.com/openshift/installer/blob/18d99d3d3e7c21d5e701036023ba5e1c137886a4/pkg/types/azure/validation/platform.go#L64

patrickdillon commented on pkg/asset/manifests/openshift.go:
```diff
@@ -271,35 +271,44 @@ func (o *Openshift) Generate(ctx context.Context, dependencies asset.Parents) er
+				aro = isAro
+			}
 		}
+		if aro && installConfig.Config.CredentialsMode != types.ManualCredentialsMode {
+			// config is used to created compatible secret to trigger azure cloud
+			// controller config merge behaviour
+			// https://github.com/openshift/origin/blob/90c050f5afb4c52ace82b15e126efe98fa798d88/vendor/k8s.io/legacy-cloud-providers/azure/azure_config.go#L83
+			session, err := installConfig.Azure.Session()
+			if err != nil {
+				return err
+			}
+			config := struct {
+				AADClientID     string `json:"aadClientId" yaml:"aadClientId"`
+				AADClientSecret string `json:"aadClientSecret" yaml:"aadClientSecret"`
+			}{
+				AADClientID:     session.Credentials.ClientID,
+				AADClientSecret: session.Credentials.ClientSecret,
+			}
```
TIL (or remembered)  ARO is generating its own secret for the cloud provider... Does this even work? These are the credentials passed to the installer, but IIUC ARO does not use a secret--they use a certificate. So I would think the value for `ClientSecret` would be empty???

The reason this would not fail during install is because the CCCMO is actually authenticating the cloud provider through its credentials request: https://github.com/openshift/cluster-cloud-controller-manager-operator/blob/master/manifests/0000_26_cloud-controller-manager-operator_14_credentialsrequest-azure.yaml

So I'm wondering if we can actually just delete this code?

rna-afk replied:
@hlipsig Do you know if there's any other specific reason we would need this or should we remove it?

patrickdillon replied:
I don't think we should resolve this until we get a better sense of whether we can actually remove the secret. https://issues.redhat.com/browse/CORS-3883 is removing the VM identity which is how the installer used to authenticate the cloud provider, but testing has shown that it is no longer needed and the cloud provider seems to be authenticated through cred requests.

hlipsig replied:
Per slack conversations we are updating this secret for ARO clusters. I don't think we have any code to create it if it's not already present so I expect it would be breaking for us.

patrickdillon replied:
> I don't think we have any code to create it if it's not already present so I expect it would be breaking for us.

The CCMO creates a credentials request which it uses to authenticate the cloud provider (see link in my first comment). We have already successfully tested removing the credentials which the installer creates for standalone openshift, so it stands to reason we can do the same for ARO. I think we should test it out so that we can drop this code.

jhixson74 replied:
Is this resolved?

patrickdillon reviewed (commented):
I would like to see if we can actually remove all ARO flags--not just the build flag. So far I haven't seen any flags we could not remove...

The only check I think we were struggling with is the one we have discussed at length, about skipping validation if the resource group is not actually empty. 

It looks like the current code is checking the `ManagedBy` field, but IIRC our conversation with @zaneb we were going to inspect tags on the resource group and skip the validation (i.e. allow non-empty resource group) if there was an ARO-specific tag (which we would need to identify) and/or a new installer specific tag.

rna-afk commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -38,8 +39,12 @@ type API interface {
 }
 
+var aro *bool
+var aroTag string = "kubernetes.io/aro"
```
Might want a better tag name

jinyunma commented on pkg/asset/installconfig/azure/client.go:
```diff
@@ -38,8 +39,12 @@ type API interface {
 }
 
+var aro *bool
+var aroTag = "kubernetes.io/aro"
```
this tag seems invalid, the character `/` is not supported.

hlipsig commented on pkg/asset/installconfig/azure/validation_test.go:
```diff
@@ -681,10 +681,9 @@ func Test_validateResourceGroup(t *testing.T) {
-		wantSkip: (&azure.Platform{}).IsARO(),
-		err:      `^\Qplatform.azure.resourceGroupName: Invalid value: "valid-resource-group-with-resources": resource group must be empty but it has 3 resources like id1, id2 ...\E$`,
+		err:       `^\Qplatform.azure.resourceGroupName: Invalid value: "valid-resource-group-with-resources": resource group must be empty but it has 3 resources like id1, id2 ...\E$`,
+	}, {
```
Why do we expect the resource group to be empty? Maybe I don't understand this code enough but most ARO customers have many things beyond the ARO cluster in the resource group. So I believe this check would be breaking.

rna-afk replied:
Fixed it. AFAIK, if the kubernetes.io tag doesn't exist or if the resource group isn't named <infraid>-rg, the installer will not blindly delete the resource group but there's no check and delete code for the inside of the resource group to see what resources the installer created and delete just those.

hlipsig commented on pkg/infrastructure/azure/azure.go:
```diff
@@ -122,20 +122,26 @@ func (p *Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionI
 		return fmt.Errorf("failed to get azure resource groups factory: %w", err)
 	}
+	// Check if the resource group exists. Create or update function replaces metadata
+	// like existing tags if the resource group exists.
```
We should not replace tags on the Resource Groups, we can append but we shouldn't do anything that remove's user tags. Once the ARO cluster is down our deny assignment will prevent new tags and this can break customer security policies if we remove tags then prevent new tag creation. ARO and Customer tags must co-locate.

rna-afk replied:
Yeah I should probably write a more clear comment. This piece of code prevents existing tag manipulation

hlipsig replied:
Great thanks!

patrickdillon commented on pkg/infrastructure/azure/azure.go:
```diff
@@ -122,20 +122,27 @@ func (p *Provider) PreProvision(ctx context.Context, in clusterapi.PreProvisionI
 	if err != nil {
 		return fmt.Errorf("failed to get azure resource groups factory: %w", err)
 	}
+	// Check if the resource group exists. Create or update function replaces metadata
+	// like existing tags if the resource group exists which is not what we want so
+	// adding a check to see if it exists and do nothing if it does.
```
I added 90f61a065e0dac2ba0f9ea54329e688a1fd643cc to  https://github.com/openshift/installer/pull/9538 which drops this code and uses CAPZ to reconcile the resource group, so I think we can drop this commit in favor of that.

rna-afk replied:
@jinyunma FYI I'll be dropping the fix in favour of Patrick's PR.

patrickdillon commented on pkg/asset/installconfig/installconfig.go:
```diff
@@ -181,15 +181,15 @@ func (a *InstallConfig) finish(ctx context.Context, filename string) error {
 		if err != nil {
 			return err
 		}
+		if aro, err := client.CheckIfARO(context.TODO(), a.Config.Azure.ResourceGroupName); err == nil && aro {
+			// ARO performs platform validation in the Resource Provider before
+			// the Installer is called. Set isARO function value to true.
+			return nil
+		}
```
Do we really need to skip validation for aro? Why? If it's validated once, it should be fine to validate again? Ideally the lofty goal would be to share common validation...

patrickdillon commented on pkg/asset/installconfig/platformprovisioncheck.go:
```diff
@@ -83,6 +84,19 @@ func (a *PlatformProvisionCheck) Generate(ctx context.Context, dependencies asse
+			}
+		}
+
+		if isAro || ic.Config.Publish != types.InternalPublishingStrategy {
+			if ic.Config.Azure.BaseDomainResourceGroupName == "" {
+				return field.Required(field.NewPath("baseDomainResourceGroupName"), "baseDomainResourceGroupName is the resource group name where the azure dns zone is deployed")
+			}
```
This doesn't seem unique to ARO? Can we just remove the isARo flag here?

rna-afk replied:
Wrong polarity. Fixed it.

patrickdillon reviewed (commented):
@rna-afk you mentioned in scrum that this is ready for tags. But I still think there are outstanding questions. I would suggest to decompose this into smaller pieces:

- we're still waiting to see whether we can remove the ARO secret for the cloud provider, I think it makes sense to spin the removal of that into its own separate PR
- is the aroTag something that has ben agreed upon with ARO? I believe the idea we floated was to look for an already existing tag used by ARO, not sure if we determinedt that existed.
Also is that tag being used to skip the validation of  resource groups containing resources? I thought that was the goal, but I missed it in the pr.

rna-afk commented:
>  we're still waiting to see whether we can remove the ARO secret for the cloud provider, I think it makes sense to spin the removal of that into its own separate PR

There's a #9506 PR that does that. I created it for testing and once this merges, it'll just remove the secret.

> is the aroTag something that has ben agreed upon with ARO? I believe the idea we floated was to look for an already existing tag used by ARO, not sure if we determined that existed.

They don't have a tag at the moment but I think if we mention the fact that a tag is needed, they will add it. I have set it to installer-aro.

> Also is that tag being used to skip the validation of resource groups containing resources? I thought that was the goal, but I missed it in the pr.
I removed the checking of resources in the resource group for all scenarios. Don't we want to install into existing resource groups even if they have resources in them? I understand it's a big functional change to the installer so I'll bring the ARO check back if we don't want this in this PR.

Removed this piece and it's supporting piece of code.
>allErrs = append(allErrs, field.Invalid(fieldPath.Child("resourceGroupName"), platform.ResourceGroupName, fmt.Sprintf("resource group must be empty but it has %d resources like %s ...", l, strings.Join(ids, ", "))))

zaneb commented on pkg/asset/machines/azure/machines.go:
```diff
@@ -213,7 +213,7 @@ func provider(platform *azure.Platform, mpool *azure.MachinePool, osImage string
 	}
 
 	managedIdentity := fmt.Sprintf("%s-identity", clusterID)
-	if platform.IsARO() || platform.CloudName == azure.StackCloud {
```
It would be helpful if each of these changes were in a separate commit, with a commit message describing why we think it is no longer needed. Without that it's impossible to review without repeating all of that work.

zaneb commented on pkg/asset/installconfig/platformprovisioncheck.go:
```diff
@@ -83,6 +84,19 @@ func (a *PlatformProvisionCheck) Generate(ctx context.Context, dependencies asse
+		}
+
+		if !isAro && ic.Config.Publish != types.InternalPublishingStrategy {
+			if ic.Config.Azure.BaseDomainResourceGroupName == "" {
```
It's not clear to me why ARO wants to be able to create an non-private cluster without a BaseDomainResourceGroupName. I think this will for sure cause problems in phase 3, so we might as well look into it now.

zaneb reviewed (commented):
> I removed the checking of resources in the resource group for all scenarios. Don't we want to install into existing resource groups even if they have resources in them?

No, this is the thing we wanted to check the tag for, so that users have to explicitly opt-in to allowing installation into a non-empty resource group (by adding the tag to it).

Ideally that'd be the _only_ thing we need to check the tag for.

rna-afk commented:
> Ideally that'd be the only thing we need to check the tag for.

Added check for another tag.

patrickdillon commented on pkg/asset/manifests/azure/cloudproviderconfig.go:
```diff
@@ -64,10 +63,6 @@ func (params CloudProviderConfig) JSON() (string, error) {
 		ExcludeMasterFromStandardLB: &excludeMasterFromStandardLB,
 	}
 
-	if params.ARO {
-		config.authConfig.UseManagedIdentityExtension = false
-	}
-
```
This sets `UseManagedIdentityExtension` to `true`, which is probably not acceptable to ARO. With [CORS-3883](https://issues.redhat.com/browse/CORS-3883) it should no longer be necessary for this to be true, so we should just be able to always set it to false, which is the subject of [CORS-3886](https://issues.redhat.com/browse/CORS-3886). 

Could you take CORS-3886 and open a PR to test setting it to always false (and removing this flag in that PR?) The default value is false so you should be able to just delete them from the config.

hlipsig replied:
What's the side effect of this always being false. Managed identities released in ARO about a week ago, right now we implement that via the wrapper, but the plan is to fully deprecate the wrapper.

patrickdillon replied:
> What's the side effect of this always being false. 

When I said "always" I was referring to there being no variance within code. Currently the value is always true for standalone because we (thought we) were auth'ing with the VM identity. The value is already always false for ARO. 


The docs for the config are here: https://cloud-provider-azure.sigs.k8s.io/install/configs/ and the relevant section:

> Managed Identity:
> 
>  For system-assigned managed identity: set useManagedIdentityExtension to true
>  For user-assigned managed identity: set useManagedIdentityExtension to true and also set userAssignedIdentityID

Right now the installer is always setting an empty `userAssignedIdentityID`:

https://github.com/openshift/installer/blob/85d542c8cb0cb50594f42ca7dc86d7e18ba8487c/pkg/asset/manifests/azure/cloudproviderconfig.go#L39-L44

#9538 removed the default VM identity (with no interruption in service).

patrickdillon replied:
Woops I accidentally submitted this comment prematurely. Still adding some context.

patrickdillon commented on pkg/asset/installconfig/azure/mock/azureclient_generated.go:
```diff
@@ -12,7 +12,7 @@ import (
 	network "github.com/Azure/azure-sdk-for-go/profiles/2020-09-01/network/mgmt/network"
 	compute "github.com/Azure/azure-sdk-for-go/profiles/latest/compute/mgmt/compute"
-	gomock "go.uber.org/mock/gomock"
+	gomock "github.com/golang/mock/gomock"
```
It looks like this change is causing issues. 

It should stick with the go.uber.org/mock package as that is the new standard. Not sure if that was updated by some automation?

rna-afk replied:
Everytime I run go generate mockgen, this creeps up.

patrickdillon replied:
Oh, I haven't hit that. If you don't want to look into it now, open a bug with the reproducer and we'll fix it.

patrickdillon commented on pkg/types/azure/platform.go:
```diff
@@ -193,7 +193,16 @@ func (p *Platform) NetworkSecurityGroupName(infraID string) string {
-// IsARO returns true if ARO-only modifications are enabled
-func (p *Platform) IsARO() bool {
-	return aro
+// GetStorageAccountName takes an infraID and generates a
```
It looks like this snuck in during a rebase

patrickdillon reviewed (commented):
this LGTM but I think some (of my) code snuck in from a rebase... I think this is because the azure stack PR got reverted, and is now trying to merge again...

/approve

zaneb commented:
/lgtm

zaneb commented:
/retest-required

zaneb commented:
/retest-required

rna-afk commented:
/retest-required

rna-afk commented:
/retest

patrickdillon commented:
/lgtm

rna-afk commented:
aws failure is due to a quota issue and not due to these changes in the PR.

/override ci/prow/e2e-aws-ovn

patrickdillon commented:
/skip

patrickdillon commented:
/override ci/prow/e2e-azure-ovn

test has already passed before. stuck in restest hell. let's see if this gets it merged.

patrickdillon commented:
/override ci/prow/e2e-aws-ovn

patrickdillon commented:
/override ci/prow/e2e-aws-ovn
